File: backend/src/main/java/com/NBE_4_5_2/Team5/domain/post/post/entity/ProductPost.java
package com.NBE_4_5_2.Team5.domain.post.post.entity;

import com.NBE_4_5_2.Team5.domain.base.entity.BaseTime;
import com.NBE_4_5_2.Team5.domain.post.category.entity.Category;
import com.NBE_4_5_2.Team5.domain.post.comment.entity.Comment;
import com.NBE_4_5_2.Team5.domain.post.post.enums.ProductStatus;
import com.NBE_4_5_2.Team5.domain.user.user.entity.User;
import com.NBE_4_5_2.Team5.global.exception.security.AuthenticationNotFoundException;
import com.NBE_4_5_2.Team5.global.exception.security.ForbiddenAccessException;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.SuperBuilder;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Getter
@SuperBuilder
@NoArgsConstructor
@AllArgsConstructor
@EntityListeners(AuditingEntityListener.class)
@EqualsAndHashCode(onlyExplicitlyIncluded = true, callSuper = false)
public class ProductPost extends BaseTime {

	@Id
	@Builder.Default
	@EqualsAndHashCode.Include
	@Column(updatable = false, nullable = false)
	private String id = "ppost-" + UUID.randomUUID();

	@Column(nullable = false)
	@Setter
	private String productName;

	@Column(nullable = false)
	@Setter
	private Integer productPrice;

	@ManyToOne(fetch = FetchType.LAZY)
	private User buyer; // 구매자(NULL이면 아직 구매 안 된 상태)

	@Column(nullable = false)
	@Setter
	private String title;

	@Column(nullable = false, columnDefinition = "TEXT")
	@Setter
	private String content;

	@Column(nullable = false, columnDefinition = "TEXT")
	@Setter
	private String image_urls; // 쉼표가 포함된 url 문자열

	//조회수
	@Column(nullable = false)
	@Builder.Default
	private Integer viewCount = 0;

	@Builder.Default
	@Column(nullable = false)
	private Integer likeCount = 0;

	@Enumerated(EnumType.STRING)
	@Column(nullable = false)
	@Setter
	@Builder.Default
	private ProductStatus status = ProductStatus.AVAILABLE;

	@Column(nullable = false)
	@Setter
	private Float latitude;

	@Column(nullable = false)
	@Setter
	private Float longitude;

	@OneToMany(mappedBy = "productPost", cascade = CascadeType.ALL, orphanRemoval = true)
	@Builder.Default
	private List<ProductCategory> productCategories = new ArrayList<>();

	@ManyToOne(fetch = FetchType.LAZY)
	private User writer;

	@OneToMany(mappedBy = "target", cascade = CascadeType.REMOVE, orphanRemoval = true)
	@Builder.Default
	private final List<Comment> commentList = new ArrayList<>();

	public static ProductPost create(User writer, String productName, Integer productPrice, String title,
		String content, String image_urls, Float latitude, Float longitude) {
		return ProductPost.builder()
			.writer(writer)
			.productName(productName)
			.productPrice(productPrice)
			.title(title)
			.content(content)
			.image_urls(image_urls)
			.latitude(latitude)
			.longitude(longitude)
			.build();
	}

	// 구매 처리 메서드
	public void setBuyer(User buyer) {
		this.buyer = buyer;
		this.status = ProductStatus.PURCHASED;
	}

	public void addCategories(List<Category> categories) {
		List<ProductCategory> productCategories = categories.stream()
			.map(category -> (ProductCategory) ProductCategory.builder()
				.productPost(this)
				.category(category)
				.build())
			.toList();

		this.productCategories.addAll(productCategories);
	}

	public void canModify(User writer) {
		if (writer == null) {
			throw new AuthenticationNotFoundException("401-1", "인증 정보가 없습니다.");
		}

		if (writer.isAdmin() || writer.equals(this.getWriter())) {
			return;
		}

		throw new ForbiddenAccessException("403-1", "자신이 작성한 글만 수정 가능합니다.");
	}

	public void canDelete(User writer) {
		if (writer == null) {
			throw new AuthenticationNotFoundException("401-1", "인증 정보가 없습니다.");
		}

		if (writer.isAdmin() || writer.equals(this.getWriter())) {
			return;
		}

		throw new ForbiddenAccessException("403-1", "자신이 작성한 글만 삭제 가능합니다.");
	}

	public boolean isAvailable() {
		return status == ProductStatus.AVAILABLE;
	}

	public void updateStatus(ProductStatus status) {
		this.status = status;
	}

	public void addComment(Comment comment) {
		commentList.add(comment);
	}

	//조회수 증가
	public void incrementViewCount() {
		this.viewCount++;
	}

	public Boolean isPurchasedBy(User loggedInUser) {
		return buyer.equals(loggedInUser);
	}
}

File: backend/src/main/java/com/NBE_4_5_2/Team5/domain/post/post/entity/ProductPost.kt
package com.NBE_4_5_2.Team5.domain.post.post.entity

import com.NBE_4_5_2.Team5.domain.base.entity.BaseTime
import com.NBE_4_5_2.Team5.domain.post.category.entity.Category
import com.NBE_4_5_2.Team5.domain.post.comment.entity.Comment
import com.NBE_4_5_2.Team5.domain.post.post.enums.ProductStatus
import com.NBE_4_5_2.Team5.domain.user.user.entity.User
import com.NBE_4_5_2.Team5.global.exception.security.AuthenticationNotFoundException
import com.NBE_4_5_2.Team5.global.exception.security.ForbiddenAccessException
import jakarta.persistence.*
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.util.*

@Entity
@EntityListeners(AuditingEntityListener::class)
class ProductPost() : BaseTime() {

    @Id
    @Column(updatable = false, nullable = false)
    val id: String = "ppost-" + UUID.randomUUID()

    @ManyToOne(fetch = FetchType.LAZY)
    lateinit var writer: User

    @Column(nullable = false)
    lateinit var productName: String

    @Column(nullable = false)
    var productPrice: Int = 0

    @Column(nullable = false)
    lateinit var title: String

    @Column(nullable = false, columnDefinition = "TEXT")
    lateinit var content: String

    @Column(name = "image_urls", nullable = false, columnDefinition = "TEXT")
    lateinit var imageUrls: String

    @ManyToOne(fetch = FetchType.LAZY)
    lateinit var buyer: User

    @Column(nullable = false)
    var viewCount: Int = 0

    @Column(nullable = false)
    var likeCount: Int = 0

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: ProductStatus = ProductStatus.AVAILABLE

    @Column(nullable = false)
    var latitude: Float = 0f

    @Column(nullable = false)
    var longitude: Float = 0f

    @OneToMany(mappedBy = "productPost", cascade = [CascadeType.ALL], orphanRemoval = true)
    val productCategories: MutableList<ProductCategory> = mutableListOf()

    @OneToMany(mappedBy = "target", cascade = [CascadeType.REMOVE], orphanRemoval = true)
    val commentList: MutableList<Comment> = mutableListOf()

    constructor(
        writer: User,
        productName: String,
        productPrice: Int,
        title: String,
        content: String,
        imageUrls: String,
        latitude: Float,
        longitude: Float
    ) : this() {
        this.writer = writer
        this.productName = productName
        this.productPrice = productPrice
        this.title = title
        this.content = content
        this.imageUrls = imageUrls
        this.latitude = latitude
        this.longitude = longitude
        setCreateDateNow();
    }

    fun addCategories(categories: List<Category>) {
        val newCategories = categories.map {
            ProductCategory.of(productPost = this, category = it)
        }
        productCategories.addAll(newCategories)
    }

    fun purchaseBy(buyer: User) {
        this.buyer = buyer
        this.status = ProductStatus.PURCHASED
    }

    fun canModify(writer: User?) {
        when {
            writer == null -> throw AuthenticationNotFoundException("401-1", "인증 정보가 없습니다.")
            writer.isAdmin() || writer == this.writer -> return
            else -> throw ForbiddenAccessException("403-1", "자신이 작성한 글만 수정 가능합니다.")
        }
    }

    fun canDelete(writer: User?) {
        when {
            writer == null -> throw AuthenticationNotFoundException("401-1", "인증 정보가 없습니다.")
            writer.isAdmin() || writer == this.writer -> return
            else -> throw ForbiddenAccessException("403-1", "자신이 작성한 글만 삭제 가능합니다.")
        }
    }

    fun isAvailable(): Boolean = status == ProductStatus.AVAILABLE

    fun updateStatus(status: ProductStatus) {
        this.status = status
    }

    fun addComment(comment: Comment) {
        commentList.add(comment)
    }

    fun incrementViewCount() {
        viewCount++
    }

    fun isPurchasedBy(user: User): Boolean {
        return buyer == user
    }
}